core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java

```diff
@@ -0,0 +1,218 @@
+        // get all kernel server
+        try {
+            List<InetSocketAddress> kernelServers =
+                    GlobalKernelReferences.nodeServer.oms.getServers(null);
+
+            // predict time optimization for replica on each kernel server and push it into sorted
+            // heap
+            for (InetSocketAddress ks : kernelServers) {
+                predictedTime.add(
+                        new AbstractMap.SimpleEntry<InetSocketAddress, Long>(
+                                ks, getOptimizedTime(replicaID, currentKernelServer, ks)));
+            }
+        } catch (RemoteException e) {
+            // TODO: Add logs for this exception
+            e.printStackTrace();
+            return currentKernelServer;
+        } catch (KernelServerNotFoundException e) {
+            // TODO: Add logs for this exception
+            e.printStackTrace();
+            return currentKernelServer;
+        }
+
+        AbstractMap.SimpleEntry<InetSocketAddress, Long> bestTimeOptimization =
+                predictedTime.first();
+        // if best time optimization is less or equal to zero then return current kernel server
+        if (bestTimeOptimization.getValue() <= 0) {
+            return currentKernelServer;
+        }
+
+        predictedTime.clear();
+        // return index 0 element
+        return bestTimeOptimization.getKey();
+    }
+
+    private long getOptimizedTime(
+            ReplicaID replicaID,
+            InetSocketAddress currentKernelServer,
+            InetSocketAddress futureKernelServer)
+            throws KernelServerNotFoundException {
+        long totalTimeOptimization = 0;
+
+        // get replica metric with respect to kernel clients
+        Map<InetSocketAddress, RPCMetric> replicaMetrics = null;
+        try {
+            replicaMetrics = microServiceManager.getMicroServiceMetric(replicaID);
+        } catch (MicroServiceNotFoundException e) {
+            e.printStackTrace();
+            return 0;
+        }
+
+        // get current kernelServer processor count
+        int currentKernelServerCPUCores =
+                serverManager.getKernelServerProcessorCount(currentKernelServer);
+        // get future kernel server CPU
+        int futureKernelServerCPUCores =
+                serverManager.getKernelServerProcessorCount(futureKernelServer);
+
+        // get node metrics
+        Map<InetSocketAddress, NodeMetric> currentKernelServerMetrics;
+        Map<InetSocketAddress, NodeMetric> futureKernelServerMetrics;
+        try {
+            currentKernelServerMetrics = serverManager.getKernelServerMetric(currentKernelServer);
+            futureKernelServerMetrics = serverManager.getKernelServerMetric(futureKernelServer);
+        } catch (KernelServerNotFoundException e) {
+            e.printStackTrace();
+            return 0;
+        }
+
+        long currentProcessingTime;
+        long currentByteExchanged;
+        long futureProcessingTime;
+        double currentIOTime;
```

Comment:
Lets avoid all double precision arithmetic. It's unnecessary.  Just record times in a small enough unit (e.g. microseconds or nanoseconds) and just use integer arithmetic.